pkg/ticker/ticker.go

```diff
@@ -0,0 +1,139 @@
+// Package ticker provides a dynamic ticker that can change its interval at runtime.
+// The ticker can be stopped gracefully and handles context-based termination.
+//
+// This package is useful for scenarios where periodic execution of a function is needed
+// and the interval might need to change dynamically based on runtime conditions.
+//
+// It also invokes a first tick immediately after the ticker starts. It's safe to use it concurrently.
+//
+// It also terminates gracefully when the context is done (return ctx.Err()) or when the stop signal is received.
+//
+// Example usage:
+//
+//	ticker := New(time.Second, func(ctx context.Context, t *Ticker) error {
+//	    resp, err := client.GetPrice(ctx)
+//	    if err != nil {
+//	        logger.Err(err).Error().Msg("failed to get price")
+//	        return nil
+//	    }
+//
+//	    observer.SetPrice(resp.GasPrice)
+//	    t.SetInterval(resp.GasPriceInterval)
+//
+//	    return nil
+//	})
+//
+//	err := ticker.Run(ctx)
+package ticker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"cosmossdk.io/errors"
+)
+
+// Ticker represents a ticker that will run a function periodically.
+// It also invokes BEFORE ticker starts.
+type Ticker struct {
+	interval   time.Duration
+	ticker     *time.Ticker
+	task       Task
+	signalChan chan struct{}
+
+	// runnerMu is a mutex to prevent double run
+	runnerMu sync.Mutex
+
+	// stateMu is a mutex to prevent concurrent SetInterval calls
+	stateMu sync.Mutex
+
+	stopped bool
+}
+
+// Task is a function that will be called by the Ticker
+type Task func(ctx context.Context, t *Ticker) error
+
+// New creates a new Ticker.
+func New(interval time.Duration, runner Task) *Ticker {
+	return &Ticker{interval: interval, task: runner}
+}
+
+// Run creates and runs a new Ticker.
+func Run(ctx context.Context, interval time.Duration, task Task) error {
+	return New(interval, task).Run(ctx)
+}
+
+// SecondsFromUint64 converts uint64 to time.Duration in seconds.
+func SecondsFromUint64(d uint64) time.Duration {
+	return time.Duration(d) * time.Second
+}
+
+// Run runs the ticker by blocking current goroutine. It also invokes BEFORE ticker starts.
+// Stops when (if any):
+// - context is done (returns ctx.Err())
+// - task returns an error or panics
+// - shutdown signal is received
+func (t *Ticker) Run(ctx context.Context) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic during ticker run: %v", r)
+		}
+	}()
+
+	// prevent concurrent runs
+	t.runnerMu.Lock()
```

Comment:
You might want to `TryLock()` instead then immediately return an error

Reply:
As far as I can tell, you never expect `Run()` to be called twice normally, this is just a guard against an unexpected scenario. In that case, I think you'd want to just return an error rather than risk a deadlock?

Reply:
Sorry, updated the comment. Yep, I think that deadlock is the way. Same way you'd have a deadlock if you call `wg.Wait()` twice for goroutines that run indefinitely unless signaled to stop